specifyweb/frontend/js_src/lib/components/FormSliders/RecordSelectorFromCollection.tsx

```diff
@@ -0,0 +1,128 @@
+import { raise } from '../Errors/Crash';
+import { relationshipIsToMany } from '../WbPlanView/mappingHelpers';
+import { RecordSelectorProps, RecordSelectorState } from './RecordSelector';
+import { useRecordSelector } from './RecordSelector';
+
+export function RecordSelectorFromCollection<SCHEMA extends AnySchema>({
+  collection,
+  relationship,
+  onAdd: handleAdd,
+  onDelete: handleDelete,
+  onSlide: handleSlide,
+  children,
+  defaultIndex = 0,
+  isInteraction,
+  ...rest
+}: Omit<
+  RecordSelectorProps<SCHEMA>,
+  | 'index'
+  | 'isDependent'
+  | 'onAdd'
+  | 'onDelete'
+  | 'records'
+  | 'relatedResource'
+  | 'table'
+> &
+  Partial<Pick<RecordSelectorProps<SCHEMA>, 'onAdd' | 'onDelete'>> & {
+    readonly collection: Collection<SCHEMA>;
+    readonly relationship: Relationship;
+    readonly defaultIndex?: number;
+    readonly children: (state: RecordSelectorState<SCHEMA>) => JSX.Element;
+    readonly isInteraction?: boolean;
+  }): JSX.Element | null {
+  const getRecords = React.useCallback(
+    (): RA<SpecifyResource<SCHEMA> | undefined> =>
+      Array.from(collection.models),
+    [collection]
+  );
+  const [records, setRecords] =
+    React.useState<RA<SpecifyResource<SCHEMA> | undefined>>(getRecords);
+
+  const isDependent = collection instanceof DependentCollection;
+  const isLazy = collection instanceof LazyCollection;
+  const isToOne =
+    !relationshipIsToMany(relationship) || relationship.type === 'zero-to-one';
+
+  // Listen for changes to collection
+  React.useEffect(
+    () =>
+      resourceOn(
+        collection,
+        'add remove destroy',
+        (): void => setRecords(getRecords),
+        true
+      ),
+    [collection, getRecords]
+  );
+
+  const [index, setIndex] = useTriggerState(Math.max(0, defaultIndex));
+
+  // Fetch records if needed
+  React.useEffect(() => {
+    /*
+     * BUG: make this more efficient (if going to the last record,
+     *   don't need to fetch all records in between)
+     */
+    if (
+      isLazy &&
+      collection.related?.isNew() !== true &&
+      !collection.isComplete() &&
+      collection.models[index] === undefined
+    )
+      collection
+        .fetch()
+        .then(() => setRecords(getRecords))
+        .catch(raise);
+  }, [collection, isLazy, getRecords, index, records.length]);
+
+  const state = useRecordSelector({
+    ...rest,
+    index,
+    table: collection.table.specifyTable,
+    records,
+    relatedResource: isDependent ? collection.related : undefined,
+    totalCount: collection._totalCount ?? records.length,
+    onAdd: (rawResources): void => {
+      const resources = isToOne ? rawResources.slice(0, 1) : rawResources;
+      if (isDependent && isToOne)
+        collection.related?.placeInSameHierarchy(resources[0]);
+      if (!isInteraction) collection.add(resources);
+      handleAdd?.(resources);
+      setIndex(
+        !isInteraction ? collection.models.length - 1 : collection.models.length
```

Comment:
Line 104 looks like a bug.
If you call collection.add(), then collection.models.length should increase. If you don't call it, length shouldn't increase - thus, you should be able to just use `collection.models.length - 1` in both cases, no?

Reply:
if we keep only `collection.models.length - 1` the right side of the slider will increase but the record display will be the first one and not second one (after clicking on the + you will have 1/2 instead of 2/2). We need this to move the display to the next record. 
This is due to the fact that we do not do `collection.add(resources)`  if in 'Interactions'.

Reply:
> if we keep only collection.models.length - 1 the right side of the slider will increase but the record display will be the first one and not second one (after clicking on the + you will have 1/2 instead of 2/2). We need this to move the display to the next record.

The thing is that until the user selects items in the interaction dialog, those records are not yet part of the subview, and so you can not setIndex to their future position - the code is not designed to handle that case and is either already broken in some ways in case of that or will break in the future - I don't feel safe doing that.

For example, see already what happens if user closes the interactions dialog:
An infinite spinner is not a proper behavior for this feature:
https://github.com/specify/specify7/assets/40512816/e07f7334-4c77-4b04-847b-2f349df9f070

Thus you either should find a way to set index only after the records are added to collection (i.e, probably add a prop to interaction dialog for that) or live with the issue where it goes to the index one before the correct one